engine/engine.go

```diff
@@ -34,115 +30,106 @@ func New(resultCache cache.Interface, valueStore store.Interface) *Engine {
+			v, found := e.resultCache.Get(mhs[i])
+
+			// If multihash found, check if value already exists in
+			// cache. Values in cache must already be in the value store, in
+			// which case there is no need to store anything new.
+			if found {
+				found = false
+				for j := range v {
+					if v[j].Equal(value) {
+						found = true
+						break
+					}
 				}
+				if found {
+					// The multihash was found and is already mapped to this
+					// value, so do not try to put it in the value store.  The
+					// value store will handle this, but at a higher cost
+					// requiring reading from disk.
+					if mhsCopy == nil {
+						// Copy-on-write
+						mhsCopy = make([]multihash.Multihash, len(mhs))
+						copy(mhsCopy, mhs)
+						mhs = mhsCopy
+					}
+					mhs[i] = mhs[len(mhs)-1]
+					mhs[len(mhs)-1] = nil
+					mhs = mhs[:len(mhs)-1]
+					continue
+				}
+				// Add this value to those already in the result cache, since
+				// the multihash was already cached.
+				e.resultCache.Put(value, mhs[i])
+				putVal = true
 			}
-			// Add this value to those already in the result cache
-			_, err := e.resultCache.Put(m, value)
-			if err != nil {
-				log.Errorw("failed to put value into result cache", "err", err)
-			}
+			i++
 		}
-	}
-
-	// If value was not in result cache, then store in value store
-	return e.valueStore.Put(m, value)
-}
-
-// PutMany stores one indexer.Value for multiple multihashes
-func (e *Engine) PutMany(mhs []multihash.Multihash, value indexer.Value) error {
-	if e.resultCache == nil {
-		return e.valueStore.PutMany(mhs, value)
-	}
-
-	for i := range mhs {
-		_, err := e.Put(mhs[i], value)
-		if err != nil {
-			return err
+		if !putVal {
+			// If there was no put to update existing metadata in the cache,
+			// then do it here.
+			e.resultCache.Put(value)
 		}
 	}
-	return nil
+	return e.valueStore.Put(value, mhs...)
 }
 
-// Remove removes a value for the specified multihash
-func (e *Engine) Remove(m multihash.Multihash, value indexer.Value) (bool, error) {
-	ok, err := e.valueStore.Remove(m, value)
+// Remove removes the specified value from multiple multihashes
+func (e *Engine) Remove(value indexer.Value, mhs ...multihash.Multihash) error {
+	// Remove first from valueStore
+	err := e.valueStore.Remove(value, mhs...)
 	if err != nil {
-		return false, err
+		return err
 	}
 
-	// If not located in valueStore or no result cache, nothing more to do
-	if !ok || e.resultCache == nil {
-		return false, nil
+	if e.resultCache == nil {
+		return nil
 	}
 
-	_, err = e.resultCache.Remove(m, value)
-	if err != nil {
-		return false, err
-	}
-	return true, nil
+	e.resultCache.Remove(value, mhs...)
+	return nil
 }
 
-// RemoveMany removes the specified value from multiple multihashes
-func (e *Engine) RemoveMany(mhs []multihash.Multihash, value indexer.Value) error {
+// RemoveProvider removes all values for specified provider.  This is used
+// when a provider is no longer indexed by the indexer.
```

Comment:
It is.  If an implementation satisfies an interface, then probably best to simply state that:
```
RemoveProvider implements indexer.Interface.
```
So that there is only one place to update documentation of behavior. I will clean this up.

Reply:
Fixed.  The general convention is that DRY applies to both code and documentation, so if a function is documented by the interface, then do not document it in the implementation.  This is especially true for implementations that are only available through an interface.